### PR TITLE
feat(DMS): add resource kafka smart connect

### PR DESCRIPTION
**Pull request by 346010762**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
   add resource kafka smart connect
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/dms/ TESTARGS='-run TestAccDmsKafkaSmartConnect_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms/ -v -run TestAccDmsKafkaSmartConnect_basic -timeout 360m -parallel 4
=== RUN   TestAccDmsKafkaSmartConnect_basic
=== PAUSE TestAccDmsKafkaSmartConnect_basic
=== CONT  TestAccDmsKafkaSmartConnect_basic
--- PASS: TestAccDmsKafkaSmartConnect_basic (1222.90s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       1222.942s
```

### CODE REVIEW

File: docs/resources/dms_kafka_smart_connect.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+* `instance_id` - (Required, String, ForceNew) Specifies the ID of the kafka instance.
+
+  Changing this parameter will create a new resource.
+
+* `storage_spec_code` - (Required, String, ForceNew) Specifies the storage specification code of the connector.
+
+* `bandwidth` - (Optional, String, ForceNew) Specifies the bandwidth of the connector.
+
+* `node_count` - (Optional, Int, ForceNew) Specifies the node count of the connector. Default to 2 and minimum is 2.
```
Comment (houpeng80) [inline docs/resources/dms_kafka_smart_connect.md]:
```suggestion
* `node_count` - (Optional, Int, ForceNew) Specifies the node count of the connector. Defaults to 2 and minimum is 2.
```

File: docs/resources/dms_kafka_smart_connect.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Attribute Reference
+
+In addition to all arguments above, the following attribute is exported:
+
+* `id` - The resource ID.
+
+## Import
+
+The kafka smart connect can be imported using the kafka `instance_id` and `connector_id` separated by a slash, e.g.
```
Comment (houpeng80) [inline docs/resources/dms_kafka_smart_connect.md]:
```suggestion
The kafka smart connect can be imported using the kafka `instance_id` and `id` separated by a slash, e.g.
```

File: docs/resources/dms_kafka_smart_connect.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+```
+$ terraform import huaweicloud_dms_kafka_smart_connect.test <instance_id>/<connector_id>
```
Comment (houpeng80) [inline docs/resources/dms_kafka_smart_connect.md]:
```suggestion
$ terraform import huaweicloud_dms_kafka_smart_connect.test <instance_id>/<id>
```

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	)
+	getKafkaSmartConnectClient, err := cfg.NewServiceClient(getKafkaSmartConnectProduct, region)
+	if err != nil {
+		return nil, fmt.Errorf("error creating DMS Client: %s", err)
+	}
+
+	// Split instance_id from resource id
+	parts := strings.Split(state.Primary.ID, "/")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid id format, must be <instance_id>/<connector_id>")
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go]:
```suggestion
		return nil, fmt.Errorf("invalid id format, must be <instance_id>/<id>")
```
and fix other places

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+		retry, err := handleMultiOperationsError(createErr)
+		return createKafkaSmartConnectResp, retry, err
+	}
+	r, err := common.RetryContextWithWaitForState(&common.RetryContextWithWaitForStateParam{
+		Ctx:          ctx,
+		RetryFunc:    retryFunc,
+		WaitFunc:     KafkaInstanceStateRefreshFunc(createKafkaSmartConnectClient, instanceID),
+		WaitTarget:   []string{"RUNNING"},
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		DelayTimeout: 10 * time.Second,
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]:
```suggestion
		DelayTimeout: 1 * time.Second,
```

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+		&getKafkaSmartConnectOpt)
+
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving DMS kafka smart connect")
+	}
+
+	_, respBodyerr := utils.FlattenResponse(getKafkaSmartConnectResp)
+	if respBodyerr != nil {
+		return diag.FromErr(respBodyerr)
+	}
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]: if the connector_id of the instance is empty of different from the id of the resource, id indicates the resource should be deleted

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+	getKafkaSmartConnectRespBody, err := utils.FlattenResponse(getKafkaSmartConnectResp)
+	if err != nil {
+		return nil, err
+	}
+	connectorId := utils.PathSearch("connector_id", getKafkaSmartConnectRespBody, nil)
+	connectorEnable := utils.PathSearch("connector_enable", getKafkaSmartConnectRespBody, false).(bool)
+	if connectorId == "" && !connectorEnable {
+		return nil, fmt.Errorf("error retrieving DMS kafka smart connect")
+	}
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go]: if connectorId is different the resource id, then it should return error

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+	d.SetId(instanceID + "/" + connectorId.(string))
+
+	// enable smart connect, need to wait for the instance status to be RUNNING so that the job could be completed.
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"EXTENDING"},
+		Target:       []string{"RUNNING"},
+		Refresh:      dmsInstanceStatusRefreshFunc(region, cfg, d),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        60 * time.Second,
+		PollInterval: 10 * time.Second,
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]:
```suggestion
		Delay:        5 * time.Second,
		PollInterval: 5 * time.Second,
```
and fix other places

File: docs/resources/dms_kafka_smart_connect.md (hunk 1 above)
Comment (346010762) [reply docs/resources/dms_kafka_smart_connect.md]: done

File: docs/resources/dms_kafka_smart_connect.md (hunk 2 above)
Comment (346010762) [reply docs/resources/dms_kafka_smart_connect.md]: done

File: docs/resources/dms_kafka_smart_connect.md (hunk 3 above)
Comment (346010762) [reply docs/resources/dms_kafka_smart_connect.md]: done

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go (hunk 4 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go (hunk 5 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+	}
+}
+
+func resourceDmsKafkaSmartConnectCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	// createKafkaSmartConnect: create DMS kafka smart connect
+	var (
+		createKafkaSmartConnectHttpUrl = "v2/{project_id}/instances/{instance_id}/connector"
+		createKafkaSmartConnectProduct = "dmsv2"
```
[the remark this answers is not in the document]
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go:76]: dmsv2 is used for retry function

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go (hunk 8 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go (hunk 6 above)
Comment (346010762) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_smart_connect.go]: done

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go (hunk 7 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_smart_connect_test.go]: done

**Pull request merged**